plugin-server/src/worker/ingestion/event-property-counter.ts

```diff
@@ -0,0 +1,178 @@
+    statsd?: StatsD
+
+    constructor(db: DB, statsd?: StatsD) {
+        this.db = db
+        this.statsd = statsd
+        this.eventPropertiesBuffer = { totalVolume: 0, uniqueVolume: 0, buffer: new Map() }
+        this.lastFlushAt = DateTime.now()
+    }
+
+    public async updateEventPropertyCounter(teamId: number, event: string, properties: Properties): Promise<void> {
+        this.updateEventPropertiesBuffer(teamId, event, properties)
+        // Flush every 2 minutes or 50k unique properties, whichever comes first.
+        // Additionally, a flush is broadcast every 1 minute from pluginServer.
+        if (
+            this.eventPropertiesBuffer.uniqueVolume > 50000 ||
+            DateTime.now().diff(this.lastFlushAt).as('seconds') > 120
+        ) {
+            await this.flush()
+        }
+    }
+
+    /** Save information about event properties into a custom buffer */
+    public updateEventPropertiesBuffer(teamId: number, event: string, properties: Record<string, any>): void {
+        const timestamp = DateTime.now().toSeconds()
+        let bufferForTeam = this.eventPropertiesBuffer.buffer.get(teamId)
+        if (!bufferForTeam) {
+            bufferForTeam = new Map()
+            this.eventPropertiesBuffer.buffer.set(teamId, bufferForTeam)
+        }
+        for (const [property, value] of Object.entries(properties)) {
+            const key = JSON.stringify([event, property])
+            let propertyBuffer = bufferForTeam.get(key)
+            if (!propertyBuffer) {
+                propertyBuffer = {
+                    createdAt: timestamp,
+                    lastSeenAt: timestamp,
+                    totalVolume: 1,
+                    propertyType: null,
+                    propertyTypeFormat: null,
+                }
+                bufferForTeam.set(key, propertyBuffer)
+                this.eventPropertiesBuffer.uniqueVolume += 1
+            } else {
+                propertyBuffer.createdAt = Math.min(timestamp, propertyBuffer.createdAt)
+                propertyBuffer.lastSeenAt = Math.max(timestamp, propertyBuffer.lastSeenAt)
+                propertyBuffer.totalVolume += 1
+            }
+            this.eventPropertiesBuffer.totalVolume += 1
+
+            propertyBuffer.propertyType =
+                typeof value === 'number'
+                    ? EventPropertyType.Number
+                    : typeof value === 'boolean'
+                    ? EventPropertyType.Boolean
+                    : typeof value === 'string'
+                    ? EventPropertyType.String
+                    : null
+            propertyBuffer.propertyTypeFormat = null
+
+            if (propertyBuffer.propertyType === EventPropertyType.String) {
+                const dateFormat = detectDateFormat(value)
+                if (dateFormat) {
+                    propertyBuffer.propertyType = EventPropertyType.DateTime
+                    propertyBuffer.propertyTypeFormat = dateFormat
+                }
+            }
```

Comment:
We won't be doing much with this data right now since the last plan was to base this off of the property definitions table (so set once for all events). Should we thus keep these fields around? I think it might be nice to have, and we will eventually get someone who wants more granularity. But... 🤔

Reply:
I think we shouldn't add them if we're not using them straight away (or we should have clear sign-posting for someone who finds them about what they are there for)